mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+        let calc_total_price = |src_amount: U256, lr_swap: &ClassicSwapData, order: &RpcOrderbookEntryV2| {
+            let src_amount = mm_number_from_u256(src_amount);
+            let order_price = MmNumber::from(order.price.rational.clone());
+            let dst_amount = MmNumber::from(lr_swap.dst_amount.as_str());
+            let order_amount = dst_amount.checked_div(&order_price)?;
+            let total_price = src_amount.checked_div(&order_amount);
+            log::debug!("select_best_swap order.coin={} lr_swap.dst_amount(wei)={} order_amount(to fill order, wei)={} total_price(with LR)={}", 
+                order.coin, lr_swap.dst_amount, order_amount.to_decimal(), total_price.clone().unwrap_or(MmNumber::from(0)).to_decimal());
+            total_price
+        };
+
+        self.inner
+            .values()
+            // filter out orders for which we did not get LR swap quotes and were not able to estimate needed source amount
+            .filter_map(|lr_data| match (lr_data.src_amount, lr_data.lr_swap_data.as_ref()) {
+                (Some(src_amount), Some(lr_swap_data)) => {
+                    Some((src_amount, lr_swap_data.clone(), lr_data.order.clone()))
+                },
+                (_, _) => None,
+            })
+            // calculate total price and filter out orders for which we could not calculate the total price
+            .filter_map(|(src_amount, lr_swap_data, order)| {
+                calc_total_price(src_amount, &lr_swap_data, &order)
+                    .map(|total_price| (lr_swap_data, order, total_price))
+            })
+            .min_by(|(_, _, price_0), (_, _, price_1)| price_0.cmp(price_1))
+            .map(|(lr_swap_data, order, price)| (lr_swap_data, order, price))
+            .ok_or(MmError::new(ApiIntegrationRpcError::BestLrSwapNotFound))
+    }
+}
+
+/// Finds the best swap path to buy order's best "UTXO" coins, including LR quotes to sell my token for the rel tokens from the orders
+/// base_amount is amount of UTXO coins user would like to buy
+pub async fn find_best_fill_ask_with_lr(
+    ctx: &MmArc,
+    user_token: Ticker,
+    orders: Vec<RpcOrderbookEntryV2>,
+    base_amount: &MmNumber,
+) -> MmResult<(ClassicSwapData, RpcOrderbookEntryV2, MmNumber), ApiIntegrationRpcError> {
+    let mut lr_data_map = LrDataMap::new_with_src_token(user_token, orders);
+    lr_data_map.update_with_contracts(ctx).await?;
+    lr_data_map.calc_destination_token_amounts(ctx, base_amount).await?;
+    lr_data_map.query_destination_token_prices(ctx).await?;
+    lr_data_map.estimate_source_token_amounts()?;
+    lr_data_map.run_lr_quotes(ctx).await?;
+
+    lr_data_map.select_best_swap()
+}
+
+/// Helper to process 1inch token cross prices data and return average price
+fn cross_prices_average(series: &CrossPricesSeries) -> MmNumber {
+    if series.is_empty() {
+        return MmNumber::from(0);
+    }
```

Comment:
i think we should filter away serieses that return no data, instead of assigning zero here.

Reply:
yes right,
I am fixing this in my follow-up code but added same in this PR as well